### PR TITLE
support 'west flash' with native_sim, add debugserver support

### LINKED ISSUE #36706: Some standard west commands should Just Work for native_posix board
**Is your enhancement proposal related to a problem? Please describe.**
Currently when building for the `native_posix` board it is not possible to use a west command to execute the resulting binary. This is inconsistent with the workflow used for other boards. While not all west commands make sense for the `native_posix` board, it should still be possible to at least invoke the native binary somehow.

**Describe the solution you'd like**
If the target board is `native_posix` then:
* Running `west flash` should execute `${BUILD_DIR}/zephyr/zephyr.elf` with no arguments
* Running `west attach` should execute `${BUILD_DIR}/zephyr/zephyr.elf` with the argument `-attach_uart`

**Describe alternatives you've considered**
Alternatively, there should be a `west execute` command that is only enabled for the `native_posix` board.

**Additional context**

[ This space intentionally left blank ]

### PR DESCRIPTION
**Pull request by nashif**
- west: add native_sim flash command
- west: runner: native_sim: add debugserver support

Fixes #36706

### CODE REVIEW

**Review by aescolar — COMMENTED** [review]
2 nits:
* About the rename, given that "native_sim" is the name of one of the boards in tree, and that this works with any other "native" (posix arch based) board in tree including native_posix, I would prefer if we did not call the runner also "native_sim". Maybe just "native" or "native_runner" or "native_run"..
* You can add the flash target to the other boards' cmake (native_posix and nrf_bsim)

**Review by aescolar — DISMISSED** [review]

**Review by aescolar — DISMISSED** [review]

**Comment by alwa-nordic** [comment]
Just my two cents: I think the old name `native_gdb` makes more sense than just `native`. From what I understand, the 'runners' are meant to be plug-able targets or environments. This particular runner is for setting up a GDB environment. This is why I named it `gdb` and not `native_posix` in the first place. One may want to define a coexisting runner for LLDB or even something with GUI like Ozone or Vscode. It makes sense to want to switch debuggers, just like one may want to switch between debugger probes.

**Comment by nashif** [comment]
> Just my two cents: I think the old name `native_gdb` makes more sense than just `native`.

This implements debugserver, debug and flash commands, independent of how you debug or flash etc, debug interfaces can be something else other than gdb, still covered by the same runner, i.e.. through command line arguments. We should not have a runner for gdb and another one for something else.. Also, look at `scripts/west_commands/runners/`, none of the existing runners is that gdb oriented in the name..

**Comment by alwa-nordic** [comment]
> can be something else other than gdb, still covered by the same runner, i.e.. through command line arguments

Sure, 'can'. But it's not. This runner is de-facto GDB-specific. Early abstraction is usually bad abstraction. E.g. You are adding a '-tui' parameter to this runner. This would make no sense in a GUI-environment.

Btw, it would be nice if you documented how to trigger the '-tui' parameter.

**Comment by alwa-nordic** [comment]
I just realized that `debugserver` is not `gdbserver`. Now I'm not sure about the naming at all any more. I think I don't understand Zephyr's runner concept.

**Comment by nashif** [comment]
> Btw, it would be nice if we documented how to trigger the '-tui' parameter.

` west debug --tui`

**Comment by alwa-nordic** [comment]
> > Btw, it would be nice if we documented how to trigger the '-tui' parameter.
> 
> ` west debug --tui`

I see you added code to make it work now. Pro tip: Don't rebase before force-pushing changes. Do an extra force-push just for the rebase. This makes the 'Compare'-feature of GitHub work much better.

**Comment by nashif** [comment]
changing assignee, most related to native targets, no changes to west

**Review by aescolar — DISMISSED** [review]

**Review by alwa-nordic — CHANGES_REQUESTED** [review]

File: scripts/west_commands/runners/native.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+DEFAULT_GDB_PORT = 3333
+
+class NativeSimBinaryRunner(ZephyrBinaryRunner):
+    """Runs the ELF binary under GDB."""
+
+    def __init__(self, cfg,
+                 tui=False,
+                 gdb_port=DEFAULT_GDB_PORT):
+        super().__init__(cfg)
+        self.tui_arg = ['-tui'] if tui else []
```
[the remark this answers is not in the document]
Comment (nashif) [reply scripts/west_commands/runners/native.py]: removed

File: scripts/west_commands/runners/native.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        if command == 'flash':
+            self.do_flash(**kwargs)
+        elif command == 'debug':
+            self.do_debug(**kwargs)
+        elif command == 'debugserver':
+            self.do_debugserver(**kwargs)
+        else:
+            assert False
+
+    def do_flash(self, **kwargs):
```
[the remark this answers is not in the document]
Comment (nashif) [reply scripts/west_commands/runners/native.py:65]: This is the "interface" afaik, looking at all the runners:

```
scripts/west_commands/runners/intel_cyclonev.py:            self.do_flash_elf(**kwargs)
scripts/west_commands/runners/intel_cyclonev.py:            self.do_flash_elf(**kwargs)
scripts/west_commands/runners/intel_cyclonev.py:    def do_flash_elf(self, **kwargs):
scripts/west_commands/runners/native.py:            self.do_flash(**kwargs)
scripts/west_commands/runners/native.py:    def do_flash(self, **kwargs):
scripts/west_commands/runners/nsim.py:            self.do_flash(**kwargs)
scripts/west_commands/runners/nsim.py:    def do_flash(self, **kwargs):
scripts/west_commands/runners/openocd.py:            self.do_flash_elf(**kwargs)
scripts/west_commands/runners/openocd.py:            self.do_flash(**kwargs)
scripts/west_commands/runners/openocd.py:    def do_flash(self, **kwargs):
scripts/west_commands/runners/openocd.py:    def do_flash_elf(self, **kwargs):
```

No harm in leaving this allowing future expansions

**Review by alwa-nordic — APPROVED** [review]

**Review by aescolar — APPROVED** [review]

**Pull request merged**